dataflow/src/main/java/org/checkerframework/dataflow/analysis/ForwardAnalysisImpl.java

```diff
@@ -0,0 +1,623 @@
+package org.checkerframework.dataflow.analysis;
+
+import com.sun.source.tree.LambdaExpressionTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.VariableTree;
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import javax.lang.model.type.TypeMirror;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.dataflow.cfg.ControlFlowGraph;
+import org.checkerframework.dataflow.cfg.UnderlyingAST;
+import org.checkerframework.dataflow.cfg.UnderlyingAST.CFGLambda;
+import org.checkerframework.dataflow.cfg.UnderlyingAST.CFGMethod;
+import org.checkerframework.dataflow.cfg.UnderlyingAST.Kind;
+import org.checkerframework.dataflow.cfg.block.Block;
+import org.checkerframework.dataflow.cfg.block.ConditionalBlock;
+import org.checkerframework.dataflow.cfg.block.ExceptionBlock;
+import org.checkerframework.dataflow.cfg.block.RegularBlock;
+import org.checkerframework.dataflow.cfg.block.SpecialBlock;
+import org.checkerframework.dataflow.cfg.node.LocalVariableNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.dataflow.cfg.node.ReturnNode;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.Pair;
+
+/**
+ * An implementation of a forward analysis to solve a org.checkerframework.dataflow problem given a
+ * control flow graph and a transfer function.
+ *
+ * @param <V> The abstract value type to be tracked by the analysis
+ * @param <S> The store type used in the analysis
+ * @param <T> The transfer function type that is used to approximated runtime behavior
+ */
+public class ForwardAnalysisImpl<
+                V extends AbstractValue<V>,
+                S extends Store<S>,
+                T extends ForwardTransferFunction<V, S>>
+        extends AbstractAnalysis<V, S, T> implements ForwardAnalysis<V, S, T> {
+
+    /**
+     * Number of times each block has been analyzed since the last time widening was applied. Null
+     * if maxCountBeforeWidening is -1, which implies widening isn't used for this analysis.
+     */
+    protected final @Nullable IdentityHashMap<Block, Integer> blockCount;
```

Comment:
I can also do it in this sprint. I was just wondering if you want this PR to be delivered to typetools very soon.

Reply:
I think we don't need widening for a backwards analysis, at all. So let's keep it out for now and then see whether there is something to optimize, with widening or otherwise.